File: bookmark.c
#include "bookmark.h"

ULONGLONG CreateBookmark() {
	return (ULONGLONG)EvtCreateBookmark(NULL);
}

ULONGLONG CreateBookmarkFromXML(char* xmlString) {
	size_t xmlWideLen= mbstowcs(NULL, xmlString, 0) + 1;
	LPWSTR lxmlString = malloc(xmlWideLen * sizeof(wchar_t));
	if (!lxmlString) {
		SetLastError(ERROR_NOT_ENOUGH_MEMORY);
		return 0;
	}

    // Convert Go string to wide characters
	mbstowcs(lxmlString, xmlString, xmlWideLen);
	return (ULONGLONG)EvtCreateBookmark(lxmlString);
}

int UpdateBookmark(ULONGLONG hBookmark, ULONGLONG hEvent) {
	return EvtUpdateBookmark((EVT_HANDLE)hBookmark, (EVT_HANDLE)hEvent);
}

char* RenderBookmark(ULONGLONG hBookmark) {
	DWORD dwUsed;
	DWORD dwProps;
	DWORD dwSize = 0;
    EvtRender(NULL, (EVT_HANDLE)hBookmark, EvtRenderBookmark, dwSize, NULL, &dwUsed, &dwProps);
    if (GetLastError() != ERROR_INSUFFICIENT_BUFFER){
    	return NULL;
    }
    dwSize = dwUsed + 1;
	LPWSTR xmlWide = malloc((dwSize) * sizeof(wchar_t));
	if (!xmlWide) {
		SetLastError(ERROR_NOT_ENOUGH_MEMORY);
		return NULL;
	}
	int renderResult = EvtRender(NULL, (EVT_HANDLE)hBookmark, EvtRenderBookmark, dwSize, xmlWide, &dwUsed, &dwProps);
    if (!renderResult) {
    	free(xmlWide);
    	return 0;
    }
    size_t xmlNarrowLen = wcstombs(NULL, xmlWide, 0) + 1;
	void* xmlNarrow = malloc(xmlNarrowLen);
	if (!xmlNarrow) {
		free(xmlWide);
		SetLastError(ERROR_NOT_ENOUGH_MEMORY);
		return NULL;
	}
	wcstombs(xmlNarrow, xmlWide, xmlNarrowLen);
	free(xmlWide);
	return xmlNarrow;
}

File: bookmark.go
// +build windows

package winlog

import (
	"syscall"
)

/* Create a new, empty bookmark. Bookmark handles must be closed with CloseEventHandle. */
func CreateBookmark() (BookmarkHandle, error) {
	bookmark, err := EvtCreateBookmark(nil)
	if err != nil {
		return 0, err
	}
	return BookmarkHandle(bookmark), nil
}

/* Create a bookmark from a XML-serialized bookmark. Bookmark handles must be closed with CloseEventHandle. */
func CreateBookmarkFromXml(xmlString string) (BookmarkHandle, error) {
	wideXmlString, err := syscall.UTF16PtrFromString(xmlString)
	if err != nil {
		return 0, err
	}
	bookmark, err := EvtCreateBookmark(wideXmlString)
	if bookmark == 0 {
		return 0, err
	}
	return BookmarkHandle(bookmark), nil
}

/* Update a bookmark to store the channel and ID of the given event */
func UpdateBookmark(bookmarkHandle BookmarkHandle, eventHandle EventHandle) error {
	return EvtUpdateBookmark(syscall.Handle(bookmarkHandle), syscall.Handle(eventHandle))
}

/* Serialize the bookmark as XML */
func RenderBookmark(bookmarkHandle BookmarkHandle) (string, error) {
	var dwUsed uint32
	var dwProps uint32
	EvtRender(0, syscall.Handle(bookmarkHandle), EvtRenderBookmark, 0, nil, &dwUsed, &dwProps)
	buf := make([]uint16, dwUsed)
	err := EvtRender(0, syscall.Handle(bookmarkHandle), EvtRenderBookmark, uint32(len(buf)), &buf[0], &dwUsed, &dwProps)
	if err != nil {
		return "", err
	}
	return syscall.UTF16ToString(buf), nil
}
